subsys/bluetooth/audio/has.c

```diff
@@ -0,0 +1,1330 @@
+/*
+ * Copyright (c) 2022 Codecoup
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+
+#include <device.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <net/buf.h>
+#include <sys/atomic.h>
+#include <sys/byteorder.h>
+#include <sys/check.h>
+#include <sys/math_extras.h>
+
+#include <bluetooth/audio/has.h>
+#include <bluetooth/bluetooth.h>
+#include <bluetooth/conn.h>
+#include <bluetooth/gatt.h>
+
+#include "has_internal.h"
+#include "../host/conn_internal.h"
+#include "../host/hci_core.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HAS)
+#define LOG_MODULE_NAME bt_has
+#include "common/log.h"
+
+#define BT_HAS_MAX_CLIENT MIN(CONFIG_BT_MAX_CONN, CONFIG_BT_MAX_PAIRED)
```

Comment:
The service is accessible for paired devices only, so it'll keep data for up to CONFIG_BT_MAX_PAIRED. The implementation keeps no context data for non-connected devices for memory efficiency.

So we maintain the paired and connected devices here only.

Reply:
Should it not have been `CONFIG_BT_MAX_PAIRED` then?

e.g. in the case of `CONFIG_BT_MAX_PAIRED=5` and `CONFIG_BT_MAX_CONN=1`, then this will be 1.

Reply:
It's as expected. Once the device is disconnected has_client is free'd.